core/src/main/scala/org/http4s/HeaderV2.scala

```diff
@@ -0,0 +1,292 @@
+          headers.collect(Function.unlift(fromRaw(_))).toNel
+      }
+  }
+}
+
+/** A collection of HTTP Headers */
+final class Headers(val headers: List[Header.Raw]) extends AnyVal {
+
+  /** TODO revise scaladoc
+    * Attempt to get a [[org.http4s.Header]] of type key.HeaderT from this collection
+    *
+    * @param key [[HeaderKey.Extractable]] that can identify the required header
+    * @return a scala.Option possibly containing the resulting header of type key.HeaderT
+    * @see [[Header]] object and get([[org.typelevel.ci.CIString]])
+    */
+  def get[A](implicit ev: Header.Select[A]): Option[ev.F[A]] =
+    ev.from(headers)
+
+  /** TODO revise scaladoc
+    * Attempt to get a [[org.http4s.Header]] from this collection of headers
+    *
+    * @param key name of the header to find
+    * @return a scala.Option possibly containing the resulting [[org.http4s.Header]]
+    */
+  def get(key: CIString): Option[NonEmptyList[Header.Raw]] = headers.filter(_.name == key).toNel
+
+  /** Make a new collection adding the specified headers, replacing existing `Single` headers.
+    *
+    * @param in multiple heteregenous headers [[Header]] to append to the new collection, see [[Header.ToRaw]]
+    * @return a new [[Headers]] containing the sum of the initial and input headers
+    */
+  def put(in: Header.ToRaw*): Headers =
+    if (in.isEmpty) this
+    else if (this.headers.isEmpty) Headers(in: _*)
+    else Headers.of(headers ++ in.toList.map(_.value))
+
+  /** Removes the `Content-Length`, `Content-Range`, `Trailer`, and
+    * `Transfer-Encoding` headers.
+    *
+    *  https://tools.ietf.org/html/rfc7231#section-3.3
+    */
+  def removePayloadHeaders: Headers =
+    Headers.of(headers.filterNot(h => Headers.PayloadHeaderKeys(h.name)))
+
+  def redactSensitive(
+      redactWhen: CIString => Boolean = Headers.SensitiveHeaders.contains): Headers =
+    Headers.of {
+      headers.map {
+        case h if redactWhen(h.name) => Header.Raw(h.name, "<REDACTED>")
+        case h => h
+      }
+    }
+
+  override def toString: String =
+    this.show
+}
+object Headers {
+  val empty = of(List.empty)
+
+  /** Creates a new Headers collection.
+    * The [[Header.ToRaw]] machinery allows the creation of Headers with
+    * variadic and heteregenous arguments, provided they are either:
+    * - A value of type `A`  which has a `Header[A]` in scope
+    * - A (name, value) pair of `String`
+    * - A `Header.Raw`
+    */
+  def apply(headers: Header.ToRaw*): Headers =
+    of(headers.toList.map(_.value))
+
+  /** Creates a new Headers collection from the headers
+    * Deduplicates non-recurring headers.
+    */
+  def of(headers: List[Header.Raw]): Headers =
+    if (headers.isEmpty) new Headers(List.empty)
+    else {
+      val acc = MutSet.empty[CIString]
+      val res = ListBuffer.empty[Header.Raw]
+
+      headers.foreach { h =>
+        if (h.recurring || h.name == "Set-Cookie" || acc.add(h.name))
```

Comment:
This needs to be a `CIString("Set-Cookie")`, right?

Reply:
yeah, good catch. It compiles, which is worrying.

Reply:
`===` would have caught it, though I've never got into the habit of using it myself.

Reply:
yep. Although I think SetCookie doesn't need to be special-cased anymore, given the Recurring + Semigroup design